plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/incrementalProcessors.kt

```diff
@@ -110,15 +124,45 @@ internal class AnnotationProcessorDependencyCollector(
     private val runtimeProcType: RuntimeProcType,
     private val warningCollector: (String) -> Unit
 ) {
-    private val generatedToSource = mutableMapOf<File, File?>()
+    private val generatedToSource = mutableMapOf<File, Pair<String?, File?>?>()
+    private val aggregatedTypes = mutableSetOf<String>()
+
     private var isFullRebuild = !runtimeProcType.isIncremental
 
-    internal fun add(createdFile: URI, originatingElements: Array<out Element?>) {
+    internal fun recordProcessingInputs(supportedAnnotationTypes: Set<String>, annotations: Set<TypeElement>, roundEnv: RoundEnvironment) {
```

Comment:
We are now tracking:
- annotations claimed by aggregating annotation processors
- annotations mentioned in sources
but it works only with main sources (not generated ones).

That limitation is what is causing https://youtrack.jetbrains.com/issue/KT-40493. In that issue an isolating AP is generating class files that are picked up in the next processing round by an aggregating annotation processor.

I think the approach in this PR can be simplified. We should implement tracking files annotated with aggregating annotations using your approach, but I'd change the following:
- track files, not types to make it more in line with the rest of incremental implementation
- if at least 1 file needs to be processed we then:
  - add all Java sources that have been processed with aggregating APs
  - for all class files that have been processed with aggregating APs only re-reprocess types (we can easily infer that from the .class name)
- also remove current tracking of annotations from the files, in order to simplify things.

WDYT?

Reply:
Discussed privately. File tracking could be fragile in case of inner classes -ambiguity on class name extraction